sysctl.8
.\" Copyright 1999, George Staikos (staikos@0wned.org)
.\" This file may be used subject to the terms and conditions of the
.\" GNU General Public License Version 2, or any later version
.\" at your option, as published by the Free Software Foundation.
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details."
.TH SYSCTL "8" "2021-03-29" "procps-ng" "System Administration"
.SH NAME
sysctl \- configure kernel parameters at runtime
.SH SYNOPSIS
.B sysctl
[\fIoptions\fR] [\fIvariable\fR[\fB=\fIvalue\fR]] [...]
.br
.B sysctl \-p
[\fIfile\fR or \fIregexp\fR] [...]
.SH DESCRIPTION
.B sysctl
is used to modify kernel parameters at runtime.  The parameters available
are those listed under /proc/sys/.  Procfs is required for
.B sysctl
support in Linux.  You can use
.B sysctl
to both read and write sysctl data.
.SH PARAMETERS
.TP
.I variable
The name of a key to read from.  An example is kernel.ostype.  The '/'
separator is also accepted in place of a '.'.
.TP
.IR  variable = value
To set a key, use the form
.IR  variable = value
where
.I variable
is the key and
.I value
is the value to set it to.  If the value contains quotes or characters
which are parsed by the shell, you may need to enclose the value in double
quotes.
.TP
\fB\-n\fR, \fB\-\-values\fR
Use this option to disable printing of the key name when printing values.
.TP
\fB\-e\fR, \fB\-\-ignore\fR
Use this option to ignore errors about unknown keys.
.TP
\fB\-N\fR, \fB\-\-names\fR
Use this option to only print the names.  It may be useful with shells that
have programmable completion.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Use this option to not display the values set to stdout.
.TP
\fB\-w\fR, \fB\-\-write\fR
Use this option when all arguments prescribe a key to be set.
.TP
\fB\-p\fR[\fIFILE\fR], \fB\-\-load\fR[=\fIFILE\fR]
Load in sysctl settings from the file specified or /etc/sysctl.conf if none
given.  Specifying \- as filename means reading data from standard input.
Using this option will mean arguments to
.B sysctl
are files, which are read in the order they are specified.
The file argument may be specified as regular expression.
.TP
\fB\-a\fR, \fB\-\-all\fR
Display all values currently available.
.TP
\fB\-\-deprecated\fR
Include deprecated parameters to
.B \-\-all
values listing.
.TP
\fB\-b\fR, \fB\-\-binary\fR
Print value without new line.
.TP
\fB\-\-system\fR
Load settings from all system configuration files. See the
.B SYSTEM FILE PRECEDENCE
section below.
.TP
\fB\-r\fR, \fB\-\-pattern\fR \fIpattern\fR
Only apply settings that match
.IR pattern .
The
.I pattern
uses extended regular expression syntax.
.TP
\fB\-A\fR
Alias of \fB\-a\fR
.TP
\fB\-d\fR
Alias of \fB\-h\fR
.TP
\fB\-f\fR
Alias of \fB\-p\fR
.TP
\fB\-X\fR
Alias of \fB\-a\fR
.TP
\fB\-o\fR
Does nothing, exists for BSD compatibility.
.TP
\fB\-x\fR
Does nothing, exists for BSD compatibility.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display help text and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version information and exit.
.SH SYSTEM FILE PRECEDENCE
When using the \fB\-\-system\fR option,
.B sysctl
will read files from directories in the following list in given
order from top to bottom. Once a file of a given filename is loaded, any
file of the same name in subsequent directories is ignored.

/etc/sysctl.d/*.conf
.br
/run/sysctl.d/*.conf
.br
/usr/local/lib/sysctl.d/*.conf
.br
/usr/lib/sysctl.d/*.conf
.br
/lib/sysctl.d/*.conf
.br
/etc/sysctl.conf

All configuration files are sorted in lexicographic order, regardless of the
directory they reside in. Configuration files can either be completely
replaced (by having a new configuration file with the same name in a
directory of higher priority) or partially replaced (by having a configuration
file that is ordered later).
.SH EXAMPLES
/sbin/sysctl \-a
.br
/sbin/sysctl \-n kernel.hostname
.br
/sbin/sysctl \-w kernel.domainname="example.com"
.br
/sbin/sysctl \-p/etc/sysctl.conf
.br
/sbin/sysctl \-a \-\-pattern forward
.br
/sbin/sysctl \-a \-\-pattern forward$
.br
/sbin/sysctl \-a \-\-pattern 'net.ipv4.conf.(eth|wlan)0.arp'
.br
/sbin/sysctl \-\-system \-\-pattern '\[char94]net.ipv6'
.SH DEPRECATED PARAMETERS
The
.B base_reachable_time
and
.B retrans_time
are deprecated.  The
.B sysctl
command does not allow changing values of these
parameters.  Users who insist to use deprecated kernel interfaces should push values
to /proc file system by other means.  For example:
.PP
echo 256 > /proc/sys/net/ipv6/neigh/eth0/base_reachable_time
.SH FILES
.I /proc/sys
.br
.I /etc/sysctl.d/*.conf
.br
.I /run/sysctl.d/*.conf
.br
.I /usr/local/lib/sysctl.d/*.conf
.br
.I /usr/lib/sysctl.d/*.conf
.br
.I /lib/sysctl.d/*.conf
.br
.I /etc/sysctl.conf
.SH SEE ALSO
.BR sysctl.conf (5)
.BR regex (7)
.SH AUTHOR
.UR staikos@0wned.org
George Staikos
.UE
.SH "REPORTING BUGS"
Please send bug reports to
.UR procps@freelists.org
.UE


NEWS
procps-ng-NEXT
---------------
  * Rename pwait to pidwait
  * library: renamed to libproc-2 and reset to 0:0:0
  * pkill: Check for lt- variants of program name          issue #192
  * ps: Add OOM and OOMADJ fields                          issue #198
  * slabtop: Don't combine d and o options                 issue #160

procps-ng-3.3.17
---------------
  * library: Incremented to 8:3:0
    (no removals or additions, internal changes only)
  * all: properly handle utf8 cmdline translations         issue #176
  * kill: Pass int to signalled process                    merge #32
  * pgrep: Pass int to signalled process                   merge #32
  * pgrep: Check sanity of SG_ARG_MAX                      issue #152
  * pgrep: Add older than selection                        merge #79
  * pidof: Quiet mode                                      merge #83
  * pidof: show worker threads                             Redhat #1803640
  * ps.1: Mention stime alias                              issue #164
  * ps: check also match on truncated 16 char comm names
  * pwait: New command waits for a process                 merge #97
  * sysctl: Match systemd directory order                  Debian #950788
  * sysctl: Document directory order                       Debian #951550
  * top: ensure config file backward compatibility         Debian #951335
  * top: add command line 'e' for symmetry with 'E'        issue #165
  * top: add '4' toggle for two abreast cpu display        issue #172
  * top: add '!' toggle for combining multiple cpus
  * top: fix potential SEGV involving -p switch            merge #114
  * vmstat: Wide mode gives wider proc columns             merge #48
  * watch: Add environment variable for interval           merge #62
  * watch: Add no linewrap option                          issue #182
  * watch: Support more colors                             merge #106,#109
  * free,uptime,slabtop: complain about extra ops          issue #181

procps-ng-3.3.16
----------------
  * docs: Use correct symbols for -h option in free.1      Debian #898774
  * docs: ps.1 now warns about command name length         issue #101
  * pgrep: Match on runstate                               issue #109, Debian #919381
  * top: can now exploit 256-color terminals               issue #96
  * top: preserves 'other filters' in configuration file   issue #99
  * top: can now collapse/expand forest view children      issue #99
  * top: parent %CPU time includes collapsed children
  * top: improve xterm support for vim navigation keys     issue #135
  * top: avoid segmentation fault at program termination   Redhat #1737552

procps-ng-3.3.15
----------------
  * library: Increment to 8:0:1
    No removals, no new functions
    Changes: slab and pid structures
  * library: Just check for SIGLOST and don't delete it    issue #93
  * library: Fix integer overflow and LPE in file2strvec   CVE-2018-1124
  * library: Use size_t for alloc functions                CVE-2018-1126
  * library: Increase comm size to 64
  * pgrep: Fix stack-based buffer overflow                 CVE-2018-1125
  * pgrep: Remove >15 warning as comm can be longer        issue #92
  * ps: Fix buffer overflow in output buffer, causing DOS  CVE-2018-1123
  * ps: Increase command name selection field to 64
  * top: Don't use cwd for location of config              CVE-2018-1122

procps-ng-3.3.14
----------------
  * free: Fix scaling on 32-bit systems                    issue #89
  * pgrep: Don't segfault on non match                     Debian #894917
  * pgrep/pkill: Revert so matches all namespaces          issue #91
  * library: Build on non-glibc systems                    issue #88

procps-ng-3.3.13
----------------
  * library: Increment to 7:0:1
    No changes, no removals
    New fuctions: numa_init, numa_max_node, numa_node_of_cpu,
      numa_uninit, xalloc_err_handler
  * doc: Document I idle state in ps.1 and top.1           Debian #886967
  * free: fix some of the SI multiples                     issue #45
  * kill: -l space between name parses correctly           Debian #854407
  * library: dont use vm_min_free on non Linux             Debian #831396
  * library: dont use SIGPWR on FreeBSD                    Debian #832148
  * library: don't strip off wchan prefixes (ps & top)     Redhat #1322111
  * pgrep: warn about 15+ char name only if -f not used
  * pgrep/pkill: only match in same namespace by default   merge #41
  * pidof: specify separator between pids                  merge #58
  * pkill: Return 0 only if we can kill process            Debian #852758
  * pmap: fix duplicate output line under '-x' option      Redhat #1374061
  * ps: avoid eip/esp address truncations                  Debian #846361
  * ps: recognizes SCHED_DEADLINE as valid CPU scheduler
  * ps: display NUMA node under which a thread ran         issue #58
  * ps: Add seconds display for cputime and time           merge #43
  * ps: Add LUID field                                     merge #57 Redhat #1518986
  * sysctl: Permit empty string for value                  issue #74
  * sysctl: Don't segv when file not available             issue #76
  * sysctl: Read and write large buffers                   merge #56
  * top: add config file support for XDG specification
  * top: eliminated minor libnuma memory leak
  * top: show fewer memory decimal places (configurable)   issue #50

sysctl.c
/*
 * Sysctl 1.01 - A utility to read and manipulate the sysctl parameters
 *
 * "Copyright 1999 George Staikos
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Changelog:
 *            v1.01:
 *                   - added -p <preload> to preload values from a file
 *            Horms:
 *                   - added -q to be quiet when modifying values
 *
 * Changes by Albert Cahalan, 2002.
 */

#include <dirent.h>
#include <errno.h>
#include <getopt.h>
#include <glob.h>
#include <libgen.h>
#include <limits.h>
#include <regex.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include "c.h"
#include "fileutils.h"
#include "nls.h"
#include "xalloc.h"
#include "procio.h"

/*
 *    Globals...
 */
static const char PROC_PATH[] = "/proc/sys/";
static const char DEFAULT_PRELOAD[] = "/etc/sysctl.conf";
static const char *DEPRECATED[] = {
	"base_reachable_time",
	"retrans_time",
	""
};
static bool IgnoreDeprecated;
static bool NameOnly;
static bool PrintName;
static bool PrintNewline;
static bool IgnoreError;
static bool Quiet;
static char *pattern;

#define LINELEN 4096
static char *iobuf;
static size_t iolen = LINELEN;

/* Function prototypes. */
static int pattern_match(const char *string, const char *pat);
static int DisplayAll(const char *restrict const path);

static void slashdot(char *restrict p, char old, char new)
{
	int warned = 1;
	p = strpbrk(p, "/.");
	if (!p)
		/* nothing -- can't be, but oh well */
		return;
	if (*p == new)
		/* already in desired format */
		return;
	while (p) {
		char c = *p;
		if ((*(p + 1) == '/' || *(p + 1) == '.') && warned) {
			xwarnx(_("separators should not be repeated: %s"), p);
			warned = 0;
		}
		if (c == old)
			*p = new;
		if (c == new)
			*p = old;
		p = strpbrk(p + 1, "/.");
	}
}

/*
 * Display the usage format
 */
static void __attribute__ ((__noreturn__))
    Usage(FILE * out)
{
	fputs(USAGE_HEADER, out);
	fprintf(out,
	      _(" %s [options] [variable[=value] ...]\n"),
		program_invocation_short_name);
	fputs(USAGE_OPTIONS, out);
	fputs(_("  -a, --all            display all variables\n"), out);
	fputs(_("  -A                   alias of -a\n"), out);
	fputs(_("  -X                   alias of -a\n"), out);
	fputs(_("      --deprecated     include deprecated parameters to listing\n"), out);
	fputs(_("  -b, --binary         print value without new line\n"), out);
	fputs(_("  -e, --ignore         ignore unknown variables errors\n"), out);
	fputs(_("  -N, --names          print variable names without values\n"), out);
	fputs(_("  -n, --values         print only values of the given variable(s)\n"), out);
	fputs(_("  -p, --load[=<file>]  read values from file\n"), out);
	fputs(_("  -f                   alias of -p\n"), out);
	fputs(_("      --system         read values from all system directories\n"), out);
	fputs(_("  -r, --pattern <expression>\n"
		"                       select setting that match expression\n"), out);
	fputs(_("  -q, --quiet          do not echo variable set\n"), out);
	fputs(_("  -w, --write          enable writing a value to variable\n"), out);
	fputs(_("  -o                   does nothing\n"), out);
	fputs(_("  -x                   does nothing\n"), out);
	fputs(_("  -d                   alias of -h\n"), out);
	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("sysctl(8)"));

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

/*
 * Strip the leading and trailing spaces from a string
 */
static char *StripLeadingAndTrailingSpaces(char *oneline)
{
	char *t;

	if (!oneline || !*oneline)
		return oneline;

	t = oneline;
	t += strlen(oneline) - 1;

	while ((*t == ' ' || *t == '\t' || *t == '\n' || *t == '\r') && t != oneline)
		*t-- = 0;

	t = oneline;

	while ((*t == ' ' || *t == '\t') && *t != 0)
		t++;

	return t;
}

/*
 * Read a sysctl setting
 */
static int ReadSetting(const char *restrict const name)
{
	int rc = 0;
	char *restrict tmpname;
	char *restrict outname;
	ssize_t rlen;
	FILE *restrict fp;
	struct stat ts;

	if (!name || !*name) {
		xwarnx(_("\"%s\" is an unknown key"), name);
		return -1;
	}

	/* used to display the output */
	outname = xstrdup(name);
	/* change / to . */
	slashdot(outname, '/', '.');

	/* used to open the file */
	tmpname = xmalloc(strlen(name) + strlen(PROC_PATH) + 2);
	strcpy(tmpname, PROC_PATH);
	strcat(tmpname, name);
	/* change . to / */
	slashdot(tmpname + strlen(PROC_PATH), '.', '/');

	/* used to display the output */
	outname = xstrdup(name);
	/* change / to . */
	slashdot(outname, '/', '.');

	if (stat(tmpname, &ts) < 0) {
		if (!IgnoreError) {
			xwarn(_("cannot stat %s"), tmpname);
			rc = -1;
		}
		goto out;
	}
	if ((ts.st_mode & S_IRUSR) == 0)
		goto out;

	if (S_ISDIR(ts.st_mode)) {
		size_t len;
		len = strlen(tmpname);
		tmpname[len] = '/';
		tmpname[len + 1] = '\0';
		rc = DisplayAll(tmpname);
		goto out;
	}

	if (pattern && !pattern_match(outname, pattern)) {
		rc = 0;
		goto out;
	}

	if (NameOnly) {
		fprintf(stdout, "%s\n", outname);
		goto out;
	}

	fp = fprocopen(tmpname, "r");

	if (!fp) {
		switch (errno) {
		case ENOENT:
			if (!IgnoreError) {
				xwarnx(_("\"%s\" is an unknown key"), outname);
				rc = -1;
			}
			break;
		case EACCES:
			xwarnx(_("permission denied on key '%s'"), outname);
			rc = -1;
			break;
		case EIO:	    /* Ignore stable_secret below /proc/sys/net/ipv6/conf */
			rc = -1;
			break;
		default:
			xwarn(_("reading key \"%s\""), outname);
			rc = -1;
			break;
		}
	} else {
		errno = 0;
		if ((rlen = getline(&iobuf, &iolen, fp)) > 0) {
			/* this loop is required, see
			 * /sbin/sysctl -a | egrep -6 dev.cdrom.info
			 */
			do {
				char *nlptr;
				if (PrintName) {
					fprintf(stdout, "%s = ", outname);
					do {
						fprintf(stdout, "%s", iobuf);
						nlptr = &iobuf[strlen(iobuf) - 1];
						/* already has the \n in it */
						if (*nlptr == '\n')
							break;
					} while ((rlen = getline(&iobuf, &iolen, fp)) > 0);
					if (*nlptr != '\n')
						putchar('\n');
				} else {
					if (!PrintNewline) {
						nlptr = strchr(iobuf, '\n');
						if (nlptr)
							*nlptr = '\0';
					}
					fprintf(stdout, "%s", iobuf);
				}
			} while ((rlen = getline(&iobuf, &iolen, fp)) > 0);
		} else {
			switch (errno) {
			case EACCES:
				xwarnx(_("permission denied on key '%s'"),
				       outname);
				rc = -1;
				break;
			case EISDIR: {
					size_t len;
					len = strlen(tmpname);
					tmpname[len] = '/';
					tmpname[len + 1] = '\0';
					fclose(fp);
					rc = DisplayAll(tmpname);
					goto out;
				}
			case EIO:	    /* Ignore stable_secret below /proc/sys/net/ipv6/conf */
				rc = -1;
				break;
			default:
				xwarnx(_("reading key \"%s\""), outname);
				rc = -1;
			case 0:
				break;
			}
		}
		fclose(fp);
	}
      out:
	free(tmpname);
	free(outname);
	return rc;
}

static int is_deprecated(char *filename)
{
	int i;
	for (i = 0; strlen(DEPRECATED[i]); i++) {
		if (strcmp(DEPRECATED[i], filename) == 0)
			return 1;
	}
	return 0;
}

/*
 * Display all the sysctl settings
 */
static int DisplayAll(const char *restrict const path)
{
	int rc = 0;
	int rc2;
	DIR *restrict dp;
	struct dirent *restrict de;
	struct stat ts;

	dp = opendir(path);

	if (!dp) {
		xwarnx(_("unable to open directory \"%s\""), path);
		rc = -1;
	} else {
		readdir(dp);	/* skip .  */
		readdir(dp);	/* skip .. */
		while ((de = readdir(dp))) {
			char *restrict tmpdir;
			if (IgnoreDeprecated && is_deprecated(de->d_name))
				continue;
			tmpdir =
			    (char *restrict) xmalloc(strlen(path) +
						     strlen(de->d_name) +
						     2);
			sprintf(tmpdir, "%s%s", path, de->d_name);
			rc2 = stat(tmpdir, &ts);
			if (rc2 != 0) {
				xwarn(_("cannot stat %s"), tmpdir);
			} else {
				if (S_ISDIR(ts.st_mode)) {
					strcat(tmpdir, "/");
					DisplayAll(tmpdir);
				} else {
					rc |=
					    ReadSetting(tmpdir +
							strlen(PROC_PATH));
				}
			}
			free(tmpdir);
		}
		closedir(dp);
	}
	return rc;
}

/*
 * Write a sysctl setting
 */
static int WriteSetting(const char *setting)
{
	int rc = 0;
	const char *name = setting;
	const char *value;
	const char *equals;
	char *tmpname;
	char *outname;
	char *last_dot;
	bool ignore_failure;

	FILE *fp;
	struct stat ts;

	if (!name)
		/* probably don't want to display this err */
		return 0;

	equals = strchr(setting, '=');

	if (!equals) {
		xwarnx(_("\"%s\" must be of the form name=value"),
		       setting);
		return -1;
	}

	/* point to the value in name=value */
	value = equals + 1;

	if (!*name || name == equals) {
		xwarnx(_("malformed setting \"%s\""), setting);
		return -2;
	}

	ignore_failure = name[0] == '-';
	if (ignore_failure)
	    name++;

	/* used to open the file */
	tmpname = xmalloc(equals - name + 1 + strlen(PROC_PATH));
	strcpy(tmpname, PROC_PATH);
	strncat(tmpname, name, (int) (equals - name));
	tmpname[equals - name + strlen(PROC_PATH)] = 0;
	/* change . to / */
	slashdot(tmpname + strlen(PROC_PATH), '.', '/');

	/* used to display the output */
	outname = xmalloc(equals - name + 1);
	strncpy(outname, name, (int) (equals - name));
	outname[equals - name] = 0;
	/* change / to . */
	slashdot(outname, '/', '.');
	last_dot = strrchr(outname, '.');
	if (last_dot != NULL && is_deprecated(last_dot + 1)) {
		xwarnx(_("%s is deprecated, value not set"), outname);
		goto out;
        }

	if (stat(tmpname, &ts) < 0) {
		if (!IgnoreError) {
			xwarn(_("cannot stat %s"), tmpname);
			rc = -1;
		}
		goto out;
	}

	if ((ts.st_mode & S_IWUSR) == 0) {
		xwarn(_("setting key \"%s\""), outname);
		goto out;
	}

	if (S_ISDIR(ts.st_mode)) {
		xwarn(_("setting key \"%s\""), outname);
		goto out;
	}

	fp = fprocopen(tmpname, "w");

	if (!fp) {
		switch (errno) {
		case ENOENT:
			if (!IgnoreError) {
				xwarnx(_("\"%s\" is an unknown key%s"), outname, (ignore_failure?_(", ignoring"):""));
				if (!ignore_failure)
				    rc = -1;
			}
			break;
		case EPERM:
		case EROFS:
		case EACCES:
			xwarnx(_("permission denied on key \"%s\"%s"), outname, (ignore_failure?_(", ignoring"):""));
			break;
		default:
			xwarn(_("setting key \"%s\"%s"), outname, (ignore_failure?_(", ignoring"):""));
			break;
		}
		if (!ignore_failure && errno != ENOENT)
		    rc = -1;
	} else {
		rc = fprintf(fp, "%s\n", value);
		if (0 < rc)
			rc = 0;
		if (close_stream(fp) != 0)
			xwarn(_("setting key \"%s\""), outname);
		else if (rc == 0 && !Quiet) {
			if (NameOnly) {
				fprintf(stdout, "%s\n", outname);
			} else {
				if (PrintName) {
					fprintf(stdout, "%s = %s\n",
						outname, value);
				} else {
					if (PrintNewline)
						fprintf(stdout, "%s\n", value);
					else
						fprintf(stdout, "%s", value);
				}
			}
		}
	}
      out:
	free(tmpname);
	free(outname);
	return rc;
}

static int pattern_match(const char *string, const char *pat)
{
	int status;
	regex_t re;

	if (regcomp(&re, pat, REG_EXTENDED | REG_NOSUB) != 0)
		return (0);
	status = regexec(&re, string, (size_t) 0, NULL, 0);
	regfree(&re);
	if (status != 0)
		return (0);
	return (1);
}

/*
 * Preload the sysctl's from the conf file.  We parse the file and then
 * reform it (strip out whitespace).
 */
static int Preload(const char *restrict const filename)
{
	FILE *fp;
	char *t;
	int n = 0;
	int rc = 0;
	ssize_t rlen;
	char *name, *value;
	glob_t globbuf;
	int globerr;
	int globflg;
	int j;

	globflg = GLOB_NOCHECK;
#ifdef GLOB_BRACE
	globflg |= GLOB_BRACE;
#endif
#ifdef GLOB_TILDE
	globflg |= GLOB_TILDE;
#else
	if (filename[0] == '~')
		xwarnx(_("GLOB_TILDE is not supported on your platform, "
			 "the tilde in \"%s\" won't be expanded."), filename);
#endif
	globerr = glob(filename, globflg, NULL, &globbuf);

	if (globerr != 0 && globerr != GLOB_NOMATCH)
		xerr(EXIT_FAILURE, _("glob failed"));

	for (j = 0; j < globbuf.gl_pathc; j++) {
		fp = (globbuf.gl_pathv[j][0] == '-' && !globbuf.gl_pathv[j][1])
		    ? stdin : fopen(globbuf.gl_pathv[j], "r");
		if (!fp) {
			xwarn(_("cannot open \"%s\""), globbuf.gl_pathv[j]);
			rc = -1;
			goto out;
		}

		while ((rlen =  getline(&iobuf, &iolen, fp)) > 0) {
			size_t offset;

			n++;

			if (rlen < 2)
				continue;

			t = StripLeadingAndTrailingSpaces(iobuf);
			if (strlen(t) < 2)
				continue;

			if (*t == '#' || *t == ';')
				continue;

			name = strtok(t, "=");
			if (!name || !*name) {
				xwarnx(_("%s(%d): invalid syntax, continuing..."),
				       globbuf.gl_pathv[j], n);
				continue;
			}

			StripLeadingAndTrailingSpaces(name);

			if (pattern && !pattern_match(name, pattern))
				continue;

			offset = strlen(name);
			memmove(&iobuf[0], name, offset);
			iobuf[offset++] = '=';

			value = strtok(NULL, "\n\r");
			if (!value || !*value) {
				xwarnx(_("%s(%d): invalid syntax, continuing..."),
				       globbuf.gl_pathv[j], n);
				continue;
			}

			while ((*value == ' ' || *value == '\t') && *value != 0)
				value++;

			/* should NameOnly affect this? */
			memmove(&iobuf[offset], value, strlen(value));
			offset += strlen(value);
			iobuf[offset] = '\0';

			rc |= WriteSetting(iobuf);
		}

		fclose(fp);
	}
out:
	return rc;
}

struct pair {
	char *name;
	char *value;
};

static int sortpairs(const void *A, const void *B)
{
	const struct pair *a = *(struct pair * const *) A;
	const struct pair *b = *(struct pair * const *) B;
	return strcmp(a->name, b->name);
}

static int PreloadSystem(void)
{
	unsigned di, i;
	const char *dirs[] = {
		"/etc/sysctl.d",
		"/run/sysctl.d",
		"/usr/local/lib/sysctl.d",
		"/usr/lib/sysctl.d",
		"/lib/sysctl.d",
	};
	struct pair **cfgs = NULL;
	unsigned ncfgs = 0;
	int rc = 0;
	struct stat ts;
	enum { nprealloc = 16 };

	for (di = 0; di < sizeof(dirs) / sizeof(dirs[0]); ++di) {
		struct dirent *de;
		DIR *dp = opendir(dirs[di]);
		if (!dp)
			continue;

		while ((de = readdir(dp))) {
			if (!strcmp(de->d_name, ".")
			    || !strcmp(de->d_name, ".."))
				continue;
			if (strlen(de->d_name) < 5
			    || strcmp(de->d_name + strlen(de->d_name) - 5, ".conf"))
				continue;
			/* check if config already known */
			for (i = 0; i < ncfgs; ++i) {
				if (cfgs && !strcmp(cfgs[i]->name, de->d_name))
					break;
			}
			if (i < ncfgs)
				/* already in */
				continue;

			if (ncfgs % nprealloc == 0)
				cfgs =
				    xrealloc(cfgs,
					     sizeof(struct pair *) * (ncfgs +
								      nprealloc));

			if (cfgs) {
				cfgs[ncfgs] =
				    xmalloc(sizeof(struct pair) +
					    strlen(de->d_name) * 2 + 2 +
					    strlen(dirs[di]) + 1);
				cfgs[ncfgs]->name =
				    (char *)cfgs[ncfgs] + sizeof(struct pair);
				strcpy(cfgs[ncfgs]->name, de->d_name);
				cfgs[ncfgs]->value =
				    (char *)cfgs[ncfgs] + sizeof(struct pair) +
				    strlen(cfgs[ncfgs]->name) + 1;
				sprintf(cfgs[ncfgs]->value, "%s/%s", dirs[di],
					de->d_name);
				ncfgs++;
			} else {
				xerrx(EXIT_FAILURE, _("internal error"));
			}

		}
		closedir(dp);
	}
	qsort(cfgs, ncfgs, sizeof(struct cfg *), sortpairs);

	for (i = 0; i < ncfgs; ++i) {
		if (!Quiet)
			printf(_("* Applying %s ...\n"), cfgs[i]->value);
		rc |= Preload(cfgs[i]->value);
	}


	if (stat(DEFAULT_PRELOAD, &ts) == 0 && S_ISREG(ts.st_mode)) {
		if (!Quiet)
			printf(_("* Applying %s ...\n"), DEFAULT_PRELOAD);
		rc |= Preload(DEFAULT_PRELOAD);
	}

	/* cleaning */
	for (i = 0; i < ncfgs; ++i) {
		free(cfgs[i]);
	}
	if (cfgs) free(cfgs);

	return rc;
}

/*
 * Main...
 */
int main(int argc, char *argv[])
{
	bool WriteMode = false;
	bool DisplayAllOpt = false;
	bool preloadfileOpt = false;
	int ReturnCode = 0;
	int c;
	const char *preloadfile = NULL;

	enum {
		DEPRECATED_OPTION = CHAR_MAX + 1,
		SYSTEM_OPTION
	};
	static const struct option longopts[] = {
		{"all", no_argument, NULL, 'a'},
		{"deprecated", no_argument, NULL, DEPRECATED_OPTION},
		{"binary", no_argument, NULL, 'b'},
		{"ignore", no_argument, NULL, 'e'},
		{"names", no_argument, NULL, 'N'},
		{"values", no_argument, NULL, 'n'},
		{"load", optional_argument, NULL, 'p'},
		{"quiet", no_argument, NULL, 'q'},
		{"write", no_argument, NULL, 'w'},
		{"system", no_argument, NULL, SYSTEM_OPTION},
		{"pattern", required_argument, NULL, 'r'},
		{"help", no_argument, NULL, 'h'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

#ifdef HAVE_PROGRAM_INVOCATION_NAME
	program_invocation_name = program_invocation_short_name;
#endif
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

	PrintName = true;
	PrintNewline = true;
	IgnoreError = false;
	Quiet = false;
	IgnoreDeprecated = true;

	if (argc < 2)
		Usage(stderr);

	while ((c =
		getopt_long(argc, argv, "bneNwfp::qoxaAXr:Vdh", longopts,
			    NULL)) != -1) {
		switch (c) {
		case 'b':
			/* This is "binary" format, which means more for BSD. */
			PrintNewline = false;
			/* FALL THROUGH */
		case 'n':
			PrintName = false;
			break;
		case 'e':
			/*
			 * For FreeBSD, -e means a "%s=%s\n" format.
			 * ("%s: %s\n" default). We (and NetBSD) use
			 * "%s = %s\n" always, and -e to ignore errors.
			 */
			IgnoreError = true;
			break;
		case 'N':
			NameOnly = true;
			break;
		case 'w':
			WriteMode = true;
			break;
		case 'f':	/* the NetBSD way */
		case 'p':
			preloadfileOpt = true;
			if (optarg)
				preloadfile = optarg;
			break;
		case 'q':
			Quiet = true;
			break;
		case 'o':	/* BSD: binary values too, 1st 16 bytes in hex */
		case 'x':	/* BSD: binary values too, whole thing in hex */
			/* does nothing */ ;
			break;
		case 'a':	/* string and integer values (for Linux, all of them) */
		case 'A':	/* same as -a -o */
		case 'X':	/* same as -a -x */
			DisplayAllOpt = true;
			break;
		case DEPRECATED_OPTION:
			IgnoreDeprecated = false;
			break;
		case SYSTEM_OPTION:
			IgnoreError = true;
			return PreloadSystem();
		case 'r':
			pattern = xstrdup(optarg);
			break;
		case 'V':
			printf(PROCPS_NG_VERSION);
			return EXIT_SUCCESS;
		case 'd':	/* BSD: print description ("vm.kvm_size: Size of KVM") */
		case 'h':	/* BSD: human-readable (did FreeBSD 5 make -e default?) */
		case '?':
			Usage(stdout);
		default:
			Usage(stderr);
		}
	}

	argc -= optind;
	argv += optind;

	iobuf = xmalloc(iolen);

	if (DisplayAllOpt)
		return DisplayAll(PROC_PATH);

	if (preloadfileOpt) {
		int ret = EXIT_SUCCESS, i;
		if (!preloadfile) {
			if (!argc) {
				ret |= Preload(DEFAULT_PRELOAD);
			}
		} else {
			/* This happens when -pfile option is
			 * used without space. */
			ret |= Preload(preloadfile);
		}
		for (i = 0; i < argc; i++)
			ret |= Preload(argv[i]);
		return ret;
	}

	if (argc < 1)
		xerrx(EXIT_FAILURE, _("no variables specified\n"
				      "Try `%s --help' for more information."),
		      program_invocation_short_name);
	if (NameOnly && Quiet)
		xerrx(EXIT_FAILURE, _("options -N and -q cannot coexist\n"
				      "Try `%s --help' for more information."),
		      program_invocation_short_name);

	for ( ; *argv; argv++) {
		if (WriteMode || strchr(*argv, '='))
			ReturnCode += WriteSetting(*argv);
		else
			ReturnCode += ReadSetting(*argv);
	}
	return ReturnCode;
}
